projects/mysql-cluster/README.md

```diff
@@ -0,0 +1,218 @@
+     GRANT REPLICATION SLAVE ON *.* TO 'replica_user'@'%';
+     FLUSH PRIVILEGES;
+     ```
+
+5. **Create a Sample Database Schema**
+   - Still in the MySQL console, create a test database and table:
+     ```sql
+     CREATE DATABASE cyfdb;
+     USE cyfdb;
+     CREATE TABLE users (
+         id INT AUTO_INCREMENT PRIMARY KEY,
+         name VARCHAR(100),
+         email VARCHAR(100)
+     );
+     ```
+
+### Task 2: Set Up the Secondary MySQL Server
+
+1. **Launch a Second EC2 Instance**
+   - Follow similar steps as for the primary instance, but tag it as `db-proj-<yourname>-replica` just for your reference.
+
+2. **Install MySQL Server on the Secondary Instance**
+   - Connect to the secondary instance using SSH.
+   - Repeat the installation steps as for the primary server.
+
+3. **Configure MySQL for Replication on the Secondary Server**
+   - Stop MySQL service:
+     ```bash
+     sudo systemctl stop mysql
+     ```
+   - Edit the MySQL configuration file:
+     ```bash
+     sudo nano /etc/mysql/mysql.conf.d/mysqld.cnf
+     ```
+   - Add the following lines at the end:
+     ```ini
+     server-id=2
+     relay-log=/var/log/mysql/mysql-relay-bin.log
+     log_bin=/var/log/mysql/mysql-bin.log
+     ```
+   - Start MySQL service:
+     ```bash
+     sudo systemctl start mysql
+     ```
+
+4. **Configure Replication**
+   - Obtain the master status on the primary server:
+     ```sql
+     SHOW MASTER STATUS;
+     ```
+   - Note down the `File` and `Position` values, also note down the private instance IP address (e.g. from your AWS console)
+   - On the secondary server, set up the replication:
+     ```sql
+     CHANGE MASTER TO
+         MASTER_HOST='<primary_server_ip>',
+         MASTER_USER='replica_user',
+         MASTER_PASSWORD='C90L6`!Doe{K',
+         MASTER_LOG_FILE='<file_name_from_master_status>',
+         MASTER_LOG_POS=<position_from_master_status>;
+     START SLAVE;
+     ```
+
+5. **Verify Replication**
+   - Check the replica status:
+     ```sql
+     SHOW REPLICA STATUS\G;
+     ```
+   - Ensure `Slave_IO_Running` and `Slave_SQL_Running` are both `Yes`.
+   - Something isn't quite right. Can you figure out how to fix this? There may be a few things that need to be fixed. Use `SHOW REPLICA STATUS\G` to find out what is wrong. Can you see the `cyfdb` database on the replica?
+
+### Task 3: Demonstrate Replication
+
+1. **Insert Data into the Primary Server**
+   - On the primary server, insert a new record:
+     ```sql
+     INSERT INTO cyfdb.users (name, email) VALUES ('John Doe', 'john@example.com');
+     ```
+
+2. **Verify Data on the Secondary Server**
+   - On the secondary server, query the table:
+     ```sql
+     SELECT * FROM testdb.users;
+     ```
+   - Verify that the data matches the primary server.
+   - Execute the following on the secondary server:
+   ```sql
+   INSERT INTO cyfdb.users (name, email) VALUES ('Jane Doe', 'jane@example.com');
+   ```
+   Did it work? If not, can you guess why?
```

Comment:
Take them through a mental model correction here. You can use the predict explain loop (copy from here https://workshops.codeyourfuture.io/#asking-questions) if that is helpful. 

The important thing from a learning perspective is to get them to describe their mental model and recognise the correction to it.

Reply:
Also are they meant to _fix_ this? The next step is to demonstrate failover, which feels like it depends on replication having successfully been running?

Reply:
It's not meant to succeed. You are not supposed to write to a replica.

Reply:
> It's not meant to succeed. You are not supposed to write to a replica.

Right, so to make this clearer, ask them to first **predict** what will happen when they try to write to a replica, then have them try to **do** it, then have them **explain** why this isn't the right thing to do.

Reply:
Done